### PR TITLE
Add Univariate Domain, Vanishing Polynomial, Lagrange Interpolation

### PR DESCRIPTION
**Pull request by tsunrise**
## Description

Add Domain, Vanishing Polynomial, Lagrange Interpolation. 

Algorithms are mainly adapted from https://github.com/alexchmit/perfect-constraints/tree/master/fractal/src/algebra. (Seems that there was a constraint error in original repo. Fixed it here)

closes: #52 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by tsunrise** [comment]
Oh no I forgot this PR completely. I will temporarily close this PR so that this will not be a blocker. 

Main issue is that there's something arithmetic error in `LagrangeCoefficient` calculation, that causes the constraints to fail (see https://github.com/arkworks-rs/r1cs-std/pull/53/files#diff-141d3454057f3f78cc6442604f6e7ed14b47542264544422512693f5d55db20aR272). still not figured out why. 

Feel free to open a PR if you want to implement similar feature, but I will try to reopen this once I figured out lol

**Comment by tsunrise** [comment]
Seems that I fixed the bug today. Can someone reopen this PR for me? Thanks @Pratyush @ValarDragon 

(or should I open another one)

**Comment by weikengchen** [comment]
Reopen.

**Comment by tsunrise** [comment]
sorry for delay, but this PR is open for review

**Comment by daira** [comment]
Is this general enough to handle uses of vanishing polynomials over Lagrange domains in any proof system, e.g. PLONK and Halo 2 as well as Fractal?

**Comment by tsunrise** [comment]
> 
> 
> Is this general enough to handle uses of vanishing polynomials over Lagrange domains in any proof system, e.g. PLONK and Halo 2 as well as Fractal?

I'm adapting those constraints from the old fractal codes. Let's ask @ValarDragon

**Comment by daira** [comment]
I edited my comment to narrow the question to R1CS, since Arkworks doesn't support PLONK-style arithmetization at all yet.

**Comment by ValarDragon** [comment]
These constraints should work for interpolation over any multiplicative coset of a Field

File: src/poly/domain/mod.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+use ark_ff::PrimeField;
+use ark_relations::r1cs::SynthesisError;
+use ark_std::vec::Vec;
+
+pub mod vanishing_poly;
+
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Debug)]
+/// Defines an evaluation domain over a prime field. The domain is a coset of size `1<<dim`.
+/// This domain is `h<g>` where g is `gen` and h is `offset`.
+pub struct EvaluationDomain<F: PrimeField> {
```
Comment (tsunrise) [inline src/poly/domain/mod.rs:13]: Shall we rename this struct to `CosetEvaluationDomain`? @ValarDragon

Comment (ValarDragon) [reply src/poly/domain/mod.rs:13]: Perhaps we should call this a `Radix2Domain`, and then change the native code to use cosets.

File: src/poly/evaluations/univariate/lagrange_interpolator.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        * Let t be the interpolation point, H be the multiplicative coset, with elements of the form h*g^i.
+        Compute each L_{i,H}(t) as Z_{H}(t) * v_i / (t- h g^i)
+        where:
+        - Z_{H}(t) = \prod_{j} (t-h*g^j) = (t^m-h^m), and
+        - v_{i} = 1 / \prod_{j \neq i} h(g^i-g^j).
+        Below we use the fact that v_{0} = 1/(m * h^(m-1)) and v_{i+1} = g * v_{i}.
+        We compute the inverse of each coefficient, and then batch invert the entire result.
+        TODO: explain deriviation more step by step (will do in THIS PR)
+        */
+        // TODO: Implement batch_inverse & mul like libiop for better efficiency (will do in next PR)
```
Comment (ValarDragon) [inline src/poly/evaluations/univariate/lagrange_interpolator.rs:75]: This is now implemented in algebra btw! (https://github.com/arkworks-rs/algebra/blob/master/ff/src/fields/mod.rs#L568)

Comment (tsunrise) [reply src/poly/evaluations/univariate/lagrange_interpolator.rs:75]: nice. Will fix that tomorrow!

File: src/poly/domain/mod.rs (hunk 1 above)
Comment (tsunrise) [reply src/poly/domain/mod.rs:13]: Sounds good

File: src/poly/evaluations/univariate/lagrange_interpolator.rs (hunk 2 above)
Comment (tsunrise) [reply src/poly/evaluations/univariate/lagrange_interpolator.rs:75]: oops. Seems that I already used `batch_inversion`. I just forgot to remove the comment haha

Comment (ValarDragon) [reply src/poly/evaluations/univariate/lagrange_interpolator.rs]: I think we can do this in a follow-up PR, I'll merge this for now!

**Pull request merged**